tests/test_cobs.cc

```diff
@@ -0,0 +1,146 @@
+#include "cobs.h"
+#include <cstring>
+
+/*
+Throw exceptions so we can catch them with gdb (catch catch).
+*/
+void throw_assert(bool condition, const char * const remark) {
+    if (condition == false) {
+        throw remark;
+    }
+}
+
+/*
+Fill a byte buffer with a given value.
+*/
+void fill(uint8_t * const b, size_t l, uint8_t v) {
+    std::memset(b, v, l);
+}
+
+/*
+Check that a byte buffer contains only a given value.
+*/
+void mem_is(uint8_t * const b, size_t l, uint8_t v) {
+    for (size_t i = 0; i < l; i++) {
+        throw_assert(b[i] == v, "");
+    }
+}
+
+int main() {
+    const uint8_t X = 0xdd; // non-zero payload
+    const uint8_t F = 0xcc; // fill
+
+    // Encode the empty packet.
+    {
+        uint8_t b1[1000]; fill(b1, sizeof(b1), X);
+        uint8_t b2[1000]; fill(b2, sizeof(b2), F);
+        uint8_t b3[1000]; fill(b3, sizeof(b3), F);
+        const size_t b1_len = 0;
+        const size_t b2_len = cobs::encode(b1, b1_len, b2, sizeof(b2));
+        mem_is(&b2[0],   1, 0);
+        mem_is(&b2[1], 999, F);
+        throw_assert(b2_len == 1, "");
+        const size_t b3_len = cobs::decode(b2, b2_len, b3, sizeof(b3));
+        mem_is(&b3[     0],              b1_len, X);
+        mem_is(&b3[b1_len], sizeof(b3) - b1_len, F);
+        throw_assert(b3_len == b1_len, "");
+    }
+
+    // Encode a packet of 1 non-zero byte.
+    {
+        uint8_t b1[1000]; fill(b1, sizeof(b1), X);
+        uint8_t b2[1000]; fill(b2, sizeof(b2), F);
+        uint8_t b3[1000]; fill(b3, sizeof(b3), F);
+        const size_t b1_len = 1;
+        const size_t b2_len = cobs::encode(b1, b1_len, b2, sizeof(b2));
+        mem_is(&b2[0],   1, 2);
+        mem_is(&b2[1],   1, X);
+        mem_is(&b2[2],   1, 0);
+        mem_is(&b2[3], 997, F);
+        throw_assert(b2_len == 3, "");
+        const size_t b3_len = cobs::decode(b2, b2_len, b3, sizeof(b3));
+        mem_is(&b3[     0],              b1_len, X);
+        mem_is(&b3[b1_len], sizeof(b3) - b1_len, F);
+        throw_assert(b3_len == b1_len, "");
+    }
+
+    // Encode a packet of 253 non-zero bytes.
+    {
+        uint8_t b1[1000]; fill(b1, sizeof(b1), X);
+        uint8_t b2[1000]; fill(b2, sizeof(b2), F);
+        uint8_t b3[1000]; fill(b3, sizeof(b3), F);
+        const size_t b1_len = 253;
+        const size_t b2_len = cobs::encode(b1, b1_len, b2, sizeof(b2));
+        mem_is(&b2[  0],   1, 254);
+        mem_is(&b2[  1], 253,   X);
+        mem_is(&b2[254],   1,   0);
+        mem_is(&b2[255], 745,   F);
+        throw_assert(b2_len == 255, "");
+        const size_t b3_len = cobs::decode(b2, b2_len, b3, sizeof(b3));
+        mem_is(&b3[     0],              b1_len, X);
+        mem_is(&b3[b1_len], sizeof(b3) - b1_len, F);
+        throw_assert(b3_len == b1_len, "");
+    }
+
+    // Encode a packet of 254 non-zero bytes. Result depends on whether or
+    // not the implementation shortens 1|0 to 0.
```

Comment:
I've changed my mind last night and now think we should use the standard encoding for interoperability and a more elegant implementation of `max_encoded_length` and `max_decoded_length`. I will update the tests afterwards.